<?xml version="1.0"?>
<robot name="twolinkbot" xmlns:xacro="http://ros.org/wiki/xacro">


  <xacro:include filename="macros.xacro"/>
  <link name="world">
  </link>

  <joint name="base_joint" type="fixed">
      <parent link="world"/>
      <child link="base_link"/>
  </joint>
  
  <link name="base_link">
    <visual>
        <origin xyz="0.045 0.05 0"/>
        <geometry>
          <mesh
          filename="file://$(find twolink_gazebo)/meshes/base_link.stl" />
        </geometry>
        <material name="grey">
            <color rgba="0.5 0.5 0.5 1.0"/>
        </material>
    </visual>
    <collision>
        <origin xyz="0.045 0.05 0"/>
        <geometry>
          <mesh
          filename="file://$(find twolink_gazebo)/meshes/base_link.stl" />
        </geometry>
    </collision>
    <xacro:inertial_cylinder mass="0.1" length="0.091" radius="0.05">
            <origin xyz="-0.0455 -0.0455 0.0455"/>
    </xacro:inertial_cylinder>
  </link>

  <!--START OF YOUR CODE-->
  <joint name="joint1" type="">

    <!--END OF YOUR CODE-->
  </joint>

  <link name="link1">
    <visual>
      <!--START OF YOUR CODE-->

      <!--END OF YOUR CODE-->
    </visual>

    <collision>
      <!--START OF YOUR CODE-->

      <!--END OF YOUR CODE-->
    </collision>

    <xacro:inertial_box mass="0.5" x="0.2" y="0.057" z="0.057">
            <origin xyz="-0.0715 0 0.03"/>
    </xacro:inertial_box>
  </link>

  <!--START OF YOUR CODE-->
  <joint name="joint2" type="">

    <!--END OF YOUR CODE-->
  </joint>

  <link name="link2">
    <visual>
      <!--START OF YOUR CODE-->

      <!--END OF YOUR CODE-->
    </visual>

    <collision>
      <!--START OF YOUR CODE-->

      <!--END OF YOUR CODE-->
    </collision>

    <xacro:inertial_box mass="0.2" x="0.15" y="0.048" z="0.006">
            <origin xyz="-0.051 0 0.003"/>
    </xacro:inertial_box>
  </link>




  <joint name="fix_joint" type="fixed">
    <parent link="link2"/>
    <child link="d435_holder"/>
    <origin xyz="-0.102 0 0.021"/>
  </joint>
  <link name="d435_holder">
    <visual>
        <origin xyz="0 0 0" rpy="0 0 3.14159"/>
        <geometry>
          <mesh
          filename="file://$(find twolink_gazebo)/meshes/d435_holder.stl" />
        </geometry>
        <material name="grey">
            <color rgba="0.5 0.5 0.5 1.0"/>
        </material>
    </visual>
    <collision>
        <origin xyz="0 0 0" rpy="0 0 3.14159"/>
        <geometry>
          <mesh
          filename="file://$(find twolink_gazebo)/meshes/d435_holder.stl" />
        </geometry>
    </collision>
  </link>
  <joint name="fix_joint2" type="fixed">
    <parent link="d435_holder"/>
    <child link="d435"/>
    <origin xyz="0 0 0.018" rpy="0 0 3.14159"/>
  </joint>
  <link name="d435">
    <visual>
        <origin xyz="0 0 -0.012" rpy="0 0 0"/>
        <geometry>
          <mesh
          filename="file://$(find twolink_gazebo)/meshes/d435.stl" />
        </geometry>
        <material name="grey">
            <color rgba="0.5 0.5 0.5 1.0"/> <!-- 회색 -->
        </material>
    </visual>
    <collision>
        <origin xyz="0 0 -0.012" rpy="0 0 0"/>
        <geometry>
          <mesh
          filename="file://$(find twolink_gazebo)/meshes/d435.stl" />
        </geometry>
    </collision>
  </link>

  <gazebo reference="link2">
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <gazebo reference="link1">
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <gazebo reference="base_link">
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <gazebo reference="d435_holder">
    <material>Gazebo/DarkGrey</material>
  </gazebo>
  <gazebo reference="d435">
    <material>Gazebo/Grey</material>
  </gazebo>
  
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="joint1">
        <command_interface name="position">
        </command_interface>
        <state_interface name="position">
            <param name="initial_value">0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
    </joint>

    <joint name="joint2">
        <command_interface name="position">
        </command_interface>
        <state_interface name="position">
            <param name="initial_value">0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
    </joint>
  </ros2_control>

  <gazebo>
      <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
          <parameters>$(find twolink_gazebo)/config/joint_controller.yaml</parameters>
      </plugin>
  </gazebo>


</robot>